### ISSUE TITLE
createElement optimisation

### ISSUE THREAD

**Issue by nfcampos** [acdlite/recompose#146]
if the signature of `createElement` was changed to `(Component) => (props, children) => ` then we could incur the cost of choosing the appropriate branch only once. This is made possible by the fact that for most uses (all `recompose` internal uses, at least) there's no use in checking for `key` in `props` as props passed to a component have `key` removed by `react`.
`createElement` would become something like 
```js
const createElement = (Component) => {
  if (isReferentiallyTransparentFunctionComponent(Component)) {
    const component = Component
    return (props, children) => {
      if (children) {
        return component({ ...props, children })
      }
      return component(props)
    }
  }

  return (props, children) => {
    if (children) {
      return <Component {...props}>{children}</Component>
    }

    return <Component {...props} />
  }
}
```

and users of `createElement`, eg. `shouldUpdate` something like this
```js
const shouldUpdate = test => BaseComponent => {
  const create = createElement(Component)
  return class extends Component {
    shouldComponentUpdate(nextProps) {
      return test(this.props, nextProps)
    }

    render() {
      return create(this.props)
    }
  }
}
```
What do you think?

**Comment by acdlite** [comment]
Neat idea, but couldn't we achieve the same thing with

```js
const shouldUpdate = test => BaseComponent => {
  const create = createElement.bind(null, Component)
  return class extends Component {
    shouldComponentUpdate(nextProps) {
      return test(this.props, nextProps)
    }

    render() {
      return create(this.props)
    }
  }
}
```

?

Or we could curry `createElement` so that both forms work. I think there's value in `createElement` being a drop-in replacement for `React.createElement()`.

**Comment by acdlite** [comment]
Oh duh obviously `.bind` won't work haha. Then I would prefer a "curried" version that supported both signatures.

**Comment by acdlite** [comment]
Actually, I don't think there's a way to "curry" `createElement` without making it incompatible with React's version. Maybe we need a separate version that we can use internally.

**Comment by acdlite** [comment]
Closed by #147
